### PR TITLE
[FLINK-25816][changelog] Refactor the logic of notifying materialization id to nested state backend

### PR DESCRIPTION
**Pull request by Myasuka**
## What is the purpose of the change

Correct the logic of FLINK-25524, and add tests to cover this case.

## Verifying this change

This change added tests and can be verified as follows:

ChangelogKeyedStateBackendTest related tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java [hunk 1]
```diff
@@ -526,18 +510,16 @@ public void notifyCheckpointAborted(long checkpointId) throws Exception {
         }
         Long materializationID = materializationIdByCheckpointId.remove(checkpointId);
         if (materializationID != null) {
-            Set<Long> checkpoints = pendingMaterializationConfirmations.get(materializationID);
-            checkpoints.remove(checkpointId);
-            if (checkpoints.isEmpty()) {
```
Comment (rkhachatryan) [inline flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:531]: The purpose of maintaining the `Set` of checkpoints was to not notfiy of abortion if any of the checkpoints is still not confirmed.
For example, there is
1. single materialization
1. two pending checkpoints using it
1. 1st get's aborted (and notified)
1. later, 2nd get's confirmed (and notified)

In (3), we'd notify of abortion and of completion in (4) - of the same materialization.
The Set would prevent this.

I don't see how the updated version addresses this. Or am I missing something?

**Review by rkhachatryan — CHANGES_REQUESTED** [review]
As I [mentioned](https://issues.apache.org/jira/browse/FLINK-25816?focusedCommentId=17482342&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17482342) in the ticket
> I think the bug confirms my previous apprehension that abortion notification adds risk and complexity while being absolutely not necessary.
 
> So I'd propose to just remove them (for the nested state backend).

And it seems the PR introduces a new bug (please see below). If so, it's one more evidence to remove this logic.

File: flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java (hunk 1 above)
Comment (Myasuka) [reply flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:531]: I think the newly added `lastCandidateAbortedMaterializationId` could prevent this and would not notify abortion in (3) and then notify completion in (4).

Comment (rkhachatryan) [reply flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:531]: How about this case:
1. materialization `M1` completes
1. checkpoint 1 using `M1` is started and aborted - `lastCandidateAbortedMaterializationId` is set to `1`
1. materialization `M2` completes
1. checkpoint `2` using `M2` is started and aborted - `lastCandidateAbortedMaterializationId` is set to `2`, nested backend notified of **abortion** of `2`
1. checkpoint `3` using `M2` is started and completed - `lastConfirmedMaterializationId` is set to `3`, nested backend notified of **completion** of `2`

Comment (Myasuka) [reply flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:531]: Actually, step 4 would not happen.
When we abort `chk-2`, the nested backend would be notified as abortion of `materialization-id-1`, and then the `lastCandidateAbortedMaterializationId` would be set as `2`.

Comment (rkhachatryan) [reply flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:531]: I see, I think you are right. 
I was confused by using `materializationID` in the 1st check and `lastCandidateAbortedMaterializationId` in the actual call.

**Comment by Myasuka** [comment]
Closing it as we choose to revert changes.

**Pull request closed without merging**